### PR TITLE
Adds OpenFaaS on AWS Fargate blog post

### PR DESCRIPTION
**Pull request by ewilde**
Resolves: #14

Signed-off-by: Edward Wilde <ewilde@gmail.com>

### CODE REVIEW

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Architecture
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Outline seems good. I would suggest starting with a bit of "Why is this needed?"

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+To run OpenFaaS in it's recommended configuration we need to schedule 5 containers: `gateway`, `nats`, `prometheus`, 
+`prometheus alert manager` and a platform provider in this case `faas-fargate`
+
+Cost per day $2.28 (note this does not include compute costs for actually running any functions!, more on that later)
+
+
+#### EKS
+[EKS Pricing](https://aws.amazon.com/fargate/pricing/)
+[EC2 Pricing](https://aws.amazon.com/ec2/pricing/)
+
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Might want to give dates when you took the prices. Also are prices a good idea since they change?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+### Cost
+An interesting aspect of running a serverless platform, how much is it going to cost you? Working this out on
+paper we see that for small numbers of functions fargate is significantly cheaper:
+
+#### Fargate
+[AWS Fargate Pricing](https://aws.amazon.com/fargate/pricing/)
+
+| Region              |[us-east-1](https://docs.aws.amazon.com/general/latest/gr/rande.html) |
+|:------------------|:------------|
+| Container specification:  |   0.25 vCPU with 512MB          |
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Is 512MB enough for things like Prometheus?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+officially supported platforms like are Kubernetes and Docker Swarm . The [community](https://github.com/openfaas/faas/blob/master/community.md) 
+has added other platforms that allow you to run OpenFaaS in more places. A good example of this is deploying OpenFaaS
+on Hashicorp's [nomad](https://www.hashicorp.com/blog/functions-as-a-service-with-nomad) application scheduler.
+
+### Complexity
+This brings us on to running OpenFaaS on AWS Fargate. Why would you want to do this? Largely this comes down to 
+complexity and cost. Perhaps you work in a team that has little or no knowledge of application schedulers like
+Kubernetes or Nomad. You may also work in a company that hasn't adopted Kubernetes and doesn't plan to in the immediate 
+future. So if you fit that in that camp, how do you leverage the advantage of serverless computing? A good choice would
+be to look at a managed serverless platform like [lambda](https://aws.amazon.com/lambda/). In this article we examine
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Probably should write *AWS Lambda* /

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
+---
+title: Faas on Fargate&#58; run OpenFaaS on AWS Fargate
+description: Edward Wilde introduces running OpenFaaS on AWS Fargate - a way to deploy and run functions without managing servers
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Can this be stronger / more specific? `a way to deploy and run functions without managing servers`

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+date: 2018-08-02
+image: /images/openfaas-on-fargate/astronomy-atmosphere-earth-220201.jpg
+categories:
+  - fargate
+author_staff_member: ed
+---
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: How about start with an intro paragraph that sums up the whole post?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+## Why run OpenFaaS on AWS Fargate
+One of the great things about OpenFaaS is it's portability. You can run OpenFaaS out of the box using 
+officially supported platforms like are Kubernetes and Docker Swarm . The [community](https://github.com/openfaas/faas/blob/master/community.md) 
+has added other platforms that allow you to run OpenFaaS in more places. A good example of this is deploying OpenFaaS
+on Hashicorp's [nomad](https://www.hashicorp.com/blog/functions-as-a-service-with-nomad) application scheduler.
+
+### Complexity
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: `Complexity` is quite negative, Kubernetes etc are our core target audience, we don't want to put people off "OpenFaaS"

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+Architecture features:
+
+* ** \[security\] ** we restrict traffic between containers using [aws security groups](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_SecurityGroups.html)
+* ** \[security\] ** isolate ingress traffic, i.e. from the internet, in a `public subnet`. Container workloads are 
+scheduled in a `private subnet`
+* ** \[discovery\] ** all workloads use [Amazon Route 53 auto naming](https://docs.aws.amazon.com/Route53/latest/APIReference/overview-service-discovery.html)
+to manage service registration / de-registration
+* ** \[logging\] ** we configure each container workload to log to [aws cloudwatch].
+
+## Running OpenFaaS on Fargate
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Could you use the active voice / verbs for titles?

`Run OpenFaaS on Fargate`

Reference: https://github.com/openfaas/openfaas.github.io#editorial--style-guidelines

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+The easiest way to run [OpenFaaS on Fargate](https://github.com/ewilde/faas-fargate) is to use the [terraform module](https://github.com/ewilde/terraform-aws-openfaas-fargate)
+I developed to deploy all the necessary components:
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: maybe you could say `my terraform module` rather than `[that] I developed`?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+ * [OpenFaaS gateway](https://github.com/openfaas/faas)
+ * [FaaS-Fargate provider](https://github.com/ewilde/faas-fargate) 
+ * [Prometheus](https://github.com/prometheus/prometheus)
+ * [Prometheus alert manager](https://github.com/prometheus/alertmanager)
+ * [NATS messaging server](https://github.com/nats-io/gnatsd)
+
+Simply run the following commands. **Costs will be incurred**:
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Can you be more specific `**Costs will be incurred**`? Maybe add it as a disclaimer instead?

I'd add a quick section on the end on how to tear-down the setup too.

File: _posts/2018-08-19-openfaas-on-fargate.md (hunk 3 above)
Comment (ewilde) [reply _posts/2018-08-19-openfaas-on-fargate.md]: Good spot, i think default prom settings are tailored for 3GB memory allocation.

I will adjust the pricing accordingly 
Ref: https://prometheus.io/docs/prometheus/1.8/storage/#memory-usage

Comment (ewilde) [reply _posts/2018-08-19-openfaas-on-fargate.md]: i've removed the cost section

File: _posts/2018-08-19-openfaas-on-fargate.md (hunk 10 above)
Comment (ewilde) [reply _posts/2018-08-19-openfaas-on-fargate.md]: added tear down section and disclaimer

File: _posts/2018-08-19-openfaas-on-fargate.md (hunk 6 above)
Comment (ewilde) [reply _posts/2018-08-19-openfaas-on-fargate.md]: added intro

`This article will explain why I've chosen to run OpenFaaS on AWS Fargate. We then move on to explore how to deploy a fully working environment and run our first function`

File: docker-compose.yml [hunk 11]
```diff
@@ -2,7 +2,7 @@ version: "3"
 services:
   openfaas-jekyll:
     command: jekyll serve --force_polling
-    image: jekyll/jekyll:latest
+    image: jekyll/jekyll:3.8.4
```
Comment (ewilde) [inline docker-compose.yml]: Latest tag doesn't build anymore:
```
docker-compose up
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

Recreating openfaasgithubio_openfaas-jekyll_1 ... done
Attaching to openfaasgithubio_openfaas-jekyll_1
openfaas-jekyll_1  | /usr/gem/bin/bundle:23:in `load': cannot load such file -- /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/exe/bundle (LoadError)
openfaas-jekyll_1  |    from /usr/gem/bin/bundle:23:in `<main>'
openfaasgithubio_openfaas-jekyll_1 exited with code 1
```

So rolled back to previous version

**Comment by ewilde** [comment]
@alexellis PTAL

**Comment by alexellis** [comment]
Hi @ewilde would you consider this post up to date and ready to publish? Can you fix the rebase conflicts too, please?

Alex

**Comment by ewilde** [comment]
> Hi @ewilde would you consider this post up to date and ready to publish? Can you fix the rebase conflicts too, please?
> 
> Alex

@alexellis yes up to date, ready to publish and rebased

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
+---
+title: Faas on Fargate&#58; run OpenFaaS on AWS Fargate
+description: Edward Wilde introduces running OpenFaaS on AWS Fargate - a simple, low cost way to deploy and run OpenFaaS functions without managing servers
+date: 2019-02-04
+image: /images/openfaas-on-fargate/astronomy-atmosphere-earth-220201.jpg
+categories:
+  - Fargate
+  - Automation
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: @ewilde can you check if we are using title case or lowercase in all the existing posts?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+| | | | '_ \ / _ \ '_ \| |_ / _` |/ _` \___ \
+| |_| | |_) |  __/ | | |  _| (_| | (_| |___) |
+ \___/| .__/ \___|_| |_|_|  \__,_|\__,_|____/
+      |_|
+
+
+Function created in folder: hello-world
+Stack file written: hello-world.yml
+```
+In the above command `-p` is my docker hub username where the image will get pushed to during the build process
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: hub -> Hub

[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: > In the above command `-p` is my docker hub username where the image will get pushed to during the build process

Let's use the full syntax `--prefix`

In the command above `--prefix` corresponds to my username on the Docker Hub. The function will be built into a Docker image and then pushed there through the `faas-cli up` command. You can also use your own self-hosted registry.

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        "arn:aws:ecs:eu-west-1:122668425727:service/nats",
+        "arn:aws:ecs:eu-west-1:122668425727:service/alertmanager"
+    ]
+}
+```
+
+## Deploying a function
+Functions are deployed in the standard OpenFaaS fashion, either using the CLI or function store website. 
+
+Before we get going it's worth telling the faas cli where our new environment is. Each command can take this as a parameter,
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: >Functions are deployed in the standard OpenFaaS fashion, either using the CLI or function store website.

> We can deploy functions to AWS Fargate the same way we would with another other OpenFaaS provider, through the CLI, UI or REST API.

[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: > Before we get going it's worth telling the faas cli where our new environment is. Each command can take this as a parameter,

> Let's configure the `faas-cli` to point at our OpenFaaS cluster.

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+experience, knowledge and goals. Perhaps you work in a team that has little or no knowledge of application schedulers like
+Kubernetes or Nomad. You may also work in a company that hasn't adopted Kubernetes and doesn't plan to in the immediate 
+future. So if you fit that in that camp, how do you leverage the advantage of serverless computing? A good choice would
+be to look at a managed serverless platform like [AWS Lambda](https://aws.amazon.com/lambda/). In this article we examine 
+another option, running OpenFaaS on Fargate, which I believe gives you some of the benefits of a managed service like 
+lambda with the flexibility of an open-source serverless platform like OpenFaaS.
+
+## Architecture
+OpenFaaS functions are built as docker images and deployed as docker containers. This approach will be very familiar to
+developers using microservices. The platform is composed of a `gateway`, `provider` and `monitoring` components. The
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: "are packaged as Docker images" "Docker containers"

[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: > This approach will be very familiar to developers using microservices.

> If you are already using microservices, you may find this approach very similar.

Let's suggest rather than tell.

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+on Hashicorp's [nomad](https://www.hashicorp.com/blog/functions-as-a-service-with-nomad) application scheduler.
+
+### Why use AWS Fargate?
+Your reasons for choosing AWS Fargate are likely to vary depending on your
+experience, knowledge and goals. Perhaps you work in a team that has little or no knowledge of application schedulers like
+Kubernetes or Nomad. You may also work in a company that hasn't adopted Kubernetes and doesn't plan to in the immediate 
+future. So if you fit that in that camp, how do you leverage the advantage of serverless computing? A good choice would
+be to look at a managed serverless platform like [AWS Lambda](https://aws.amazon.com/lambda/). In this article we examine 
+another option, running OpenFaaS on Fargate, which I believe gives you some of the benefits of a managed service like 
+lambda with the flexibility of an open-source serverless platform like OpenFaaS.
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: "on Fargate" -> "on AWS Fargate"

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+
+_Experience level:_
+
+_beginner OpenFaaS / intermediate AWS Fargate_
+
+This article will explain why I've chosen to run OpenFaaS on AWS Fargate. We then move on to explore
+how to deploy a fully working environment and run our first function.
+
+One of the great things about OpenFaaS is it's portability. You can run OpenFaaS out of the box using 
+officially supported platforms like are Kubernetes and Docker Swarm . The [community](https://github.com/openfaas/faas/blob/master/community.md)
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: > One of the great things about OpenFaaS is it's portability.

> For me, one of the best things about OpenFaaS is how portable it is.

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+DNS, load-balancing and service discovery
+
+_Experience level:_
+
+_beginner OpenFaaS / intermediate AWS Fargate_
+
+This article will explain why I've chosen to run OpenFaaS on AWS Fargate. We then move on to explore
+how to deploy a fully working environment and run our first function.
+
+One of the great things about OpenFaaS is it's portability. You can run OpenFaaS out of the box using
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: > why I've chosen to run OpenFaaS on AWS Fargate.

Are you going to mention that you also built the provider in order to be able to run OpenFaaS on top of AWS Fargate at Form3? I think that's an important detail

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+
+Simply run the following commands:
+
+> Disclaimer: as you would expect creating resources in AWS will incur charges. Please follow the tear-down instructions
+below when you have finished.
+
+
+### Mac
+```shell
+brew install terraform # or upgrade
```
[the remark this answers is not in the document]
Comment (alexellis) [reply _posts/2018-08-19-openfaas-on-fargate.md]: 👍

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+[0] worker done.
+
+Deploying: hello-world.
+
+Deployed. 202 Accepted.
+URL: https://openfaas.{your-domain}.com/function/hello-world
+```
+
+### Removing the installation
+* Remove any functions that you have deployed using `faas-cli remove`
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: > Firstly, remove any functions that you have deployed using `faas-cli remove FUNCTION_NAME`.

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+* Run `make destroy`. This can take between 5-10 minutes
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Briefly summarise what is happening?

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+aws_subnet.external[2]: Destruction complete after 1m48s
+aws_security_group.alb: Still destroying... (ID: sg-0a84a6bd951bacb6d, 1m50s elapsed)
+aws_security_group.alb: Destruction complete after 1m51s
+aws_vpc.default: Destroying... (ID: vpc-0220284dd7ba01f58)
+aws_vpc.default: Destruction complete after 0s
+
+Destroy complete! Resources: 133 destroyed.
+```
+
+## What's next for OpenFaaS on Fargate
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: This is a very sudden close to the post.

Could you add a conclusion section with 

* A summary of what you did
* What's next for faas-fargate and link to the repo. Mention if you welcome contributions?
* See the wrapping-up section of the last few official blogs on how to construct a call to action pointing back to the project Slack, Twitter and etc

File: _posts/2018-08-19-openfaas-on-fargate.md [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+* Support _scale to zero_ using the [faas-idler](https://github.com/openfaas-incubator/faas-idler).
+* Provide better _observability_ for users and integrate grafana with the standard deployment.
```
Comment (alexellis) [inline _posts/2018-08-19-openfaas-on-fargate.md]: Provide better observability How?  `grafana` -> Grafana.

File: docker-compose.yml (hunk 11 above)
Comment (alexellis) [reply docker-compose.yml]: I'm trying the latest tag on my Intel NUC now to verify.

Comment (ewilde) [reply docker-compose.yml]: looks like 3.8.5 has fixed this problem

File: _posts/2018-08-19-openfaas-on-fargate.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ewilde) [reply _posts/2018-08-19-openfaas-on-fargate.md]: lowercase
